packages/hyperion-autologging/src/ALNetworkPublisher.ts

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+ */
+
+import { assert } from "@hyperion/global";
+import { Channel } from "@hyperion/hook/src/Channel";
+import * as intercept from "@hyperion/hyperion-core/src/intercept";
+import * as IWindow from "@hyperion/hyperion-dom/src/IWindow";
+import * as IXMLHttpRequest from "@hyperion/hyperion-dom/src/IXMLHttpRequest";
+import performanceAbsoluteNow from "@hyperion/hyperion-util/src/performanceAbsoluteNow";
+import * as Types from "@hyperion/hyperion-util/src/Types";
+import { ALFlowletEvent, ALSharedInitOptions, ALTimedEvent } from "./ALType";
+
+type ALNetworkEvent = ALTimedEvent & Readonly<{
+  event: "network";
+  initiatorType: "fetch" | "xmlhttprequest"; // https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming/initiatorType
+  flowlet: ALFlowletEvent['flowlet'] | null;
+}>;
+
+
+/**
+ * This is a generic event to be fired when a network request is about to happen.
+ * Generally, various network api can be intercepted to provide a more unified way
+ * of notifying the app about network requests.
+ */
+type RequestInfo = {
+  method: string;
+  url: string | URL; // TODO: should we always send on of the types?
+  body?: RequestInit['body'],
+}
+export type ALNetworkRequestEvent = ALNetworkEvent & Readonly<RequestInfo>;
+
+export type ALNetworkResponseEvent = ALNetworkEvent & Readonly<
+  {
+    requestEvent: ALNetworkRequestEvent | undefined | null;
+  }
+  &
+  (
+    {
+      initiatorType: "fetch";
+      response: Response;
+    }
+    |
+    {
+      initiatorType: "xmlhttprequest";
+      response: XMLHttpRequest;
+    }
+  )
+>;
+
+export type ALChannelNetworkEvent = Readonly<{
+  al_network_request: [ALNetworkRequestEvent];
+  al_network_response: [ALNetworkResponseEvent];
+}>;
+
+
+type ALChannel = Channel<ALChannelNetworkEvent>;
+
+export type InitOptions = Types.Options<
+  ALSharedInitOptions &
+  {
+    channel: ALChannel;
+    /**
+     * if provided, only requests that pass the filter function
+     * will generate request/response events. 
+     */
+    requestFilter?: (resuest: RequestInfo) => boolean;
+  }
+>;
+
+
+function captureFetch(options: InitOptions): void {
+  const { channel, flowletManager } = options;
+
+  let requestEvent: ALNetworkResponseEvent['requestEvent'];
+  IWindow.fetch.onArgsObserverAdd((input, init) => {
+    let request: RequestInfo;
+    if (typeof input === "string") {
+      request = {
+        body: init?.body,
+        method: init?.method ?? "get",
+        url: input,
+      };
+    } else if (input instanceof Request) {
+      request = {
+        body: input.body,
+        method: input.method,
+        url: input.url,
+      };
+
+    } else {
+      request = {
+        method: "GET",
```

Comment:
We use `?? "get"` above, but `"GET"` here as fallback,  should we use the same casing or lower/upper all `method` values for consistency?